mqtt/mqtt-edgehub/src/translation.rs

```diff
@@ -0,0 +1,361 @@
+                        }
+                    )*
+                }
+
+                None
+            }
+
+            fn translate_to_old(&self, topic: &str) -> Option<String> {
+                if topic.starts_with("$edgehub/") {
+                    $(
+                        if let Some(captures) = self.to_old.$translate_name.captures(topic) {
+                            return Some($old_to(captures).into());
+                        }
+                    )*
+                }
+
+                None
+            }
+        }
+    };
+}
+
+translate_d2c! {
+    // Message Translation
+    device_send_message { // note this may have to be split into 2 patterns for device and modules, depending on how client_id encodes device and module id
+        to_new {
+            format!("devices/{}/messages/events(?P<path>.*)", DEVICE_ID),
+            {|captures: regex::Captures<'_>, _| format!("$edgehub/{}/messages/events{}", &captures["device_id"], &captures["path"])}
+        }
+    },
+
+    // Twin Translation
+    twin_send_update_to_hub {
+        to_new {
+            "\\$iothub/twin/PATCH/properties/reported/(?P<params>.*)",
+            {|captures: regex::Captures<'_>, client_id| format!("$edgehub/{}/twin/reported/{}", client_id, &captures["params"])}
+        }
+    },
+    twin_get_from_hub {
+        to_new {
+            "\\$iothub/twin/GET/(?P<params>.*)",
+            {|captures: regex::Captures<'_>, client_id| format!("$edgehub/{}/twin/get/{}", client_id, &captures["params"])}
+        }
+    },
+
+    // Direct Methods
+    direct_method_response {
+        to_new {
+            "\\$iothub/methods/res/(?P<status>.*)",
+            {|captures: regex::Captures<'_>, client_id| format!("$edgehub/{}/methods/res/{}", client_id, &captures["status"])}
+        }
+    }
+}
+
+translate_c2d! {
+    // Message Translation
+    c2d_message {
+        to_new {
+            format!("devices/{}/messages/devicebound(?P<path>.*)", DEVICE_ID),
+            {|captures: regex::Captures<'_>, _| format!("$edgehub/{}/messages/c2d/post{}", &captures["device_id"], &captures["path"])}
+        },
+        to_old {
+            format!("\\$edgehub/{}/messages/c2d/post(?P<path>.*)", DEVICE_ID),
+            {|captures: regex::Captures<'_>| format!("devices/{}/messages/devicebound{}", &captures["device_id"], &captures["path"])}
+        }
+    },
+
+    // Twin Translation
+    twin_receive_update_from_hub {
+        to_new {
+            "\\$iothub/twin/PATCH/properties/desired/(?P<params>.*)",
+            {|captures: regex::Captures<'_>, client_id| format!("$edgehub/{}/twin/desired/{}", client_id, &captures["params"])}
+        },
+        to_old {
+            format!("\\$edgehub/{}/twin/desired/(?P<params>.*)", DEVICE_ID),
+            {|captures: regex::Captures<'_>| format!("$iothub/twin/PATCH/properties/desired/{}", &captures["params"])}
+        }
+    },
+    twin_response_from_hub {
```

Comment:
missed comment?

Reply:
nope, its part of the "Twin Translation" group.

Reply:
// Message Translation
c2d_message

// Twin Translation
twin_receive_update_from_hub 
twin_response_from_hub

// Direct Methods
receive_direct_method_request